robottelo/orm.py

```diff
@@ -166,13 +169,33 @@ class attributes of ``Meta`` are metadata. Here is one way to instantiate
 
         User(name='Alice', supervisor=1, subordinate=[3, 4])
 
+    An entity object is useless if you are unable to use it to communicate with
+    a server. The solution is to provide a ``nailgun.config.ServerConfig`` when
+    instantiating a new entity. This configuration object is stored as an
+    instance variable named ``server_config_`` and used by methods such as
+    :meth:`robottelo.orm.Entity.path`.
+
+    1. If the ``server_config`` argument is specified, then that is used.
+    2. Otherwise, if :data:`robottelo.orm.DEFAULT_SERVER_CONFIG` is set, then
+       that is used.
+    3. Otherwise, call ``nailgun.config.ServerConfig.get()``.
+
     """
     # The id() builtin is still available within instance methods, class
     # methods, static methods, inner classes, and so on. However, id() is *not*
     # available at the current level of lexical scoping after this point.
     id = IntegerField()  # pylint:disable=C0103
 
-    def __init__(self, **kwargs):
+    def __init__(self, server_config=None, **kwargs):
+        # server_config > DEFAULT_SERVER_CONFIG > ServerConfig.get()
+        if server_config is not None:
+            self.server_config_ = server_config
```

Comment:
This is done to prevent naming conflicts. What if an entity has an attribute named "server_config"? In that case, the name of the attribute would conflict with this piece of instance data.

So far, the entity classes have been designed with the assumption that each instance attribute on an entity object corresponds to some field on the entity class. In other words, if you ran this code:

```python
>>> vars(server_group)
{
    'name': 'my-server-group',
    'description': 'a collection of servers that all have the same configuration',
    'server_config': <orm.ServerConfig object at ...>,
}
```

Then you could safely make the assumption that this class definition exists somewhere:

```python
class ServerGroup(orm.Entity):
    name = StringField(...)
    description = StringField(...)
    server_config = OneToOneField('ServerConfig')
```

However, that is no longer true. When you run `vars`, you'll now get this:

```python
>>> vars(server_group)
{
    'name': 'my-server-group',
    'description': 'a collection of servers that all have the same configuration',
    'server_config': <orm.ServerConfig object at ...>,
    'server_config_': <nailgun.ServerConfig object at ...>,
}
```

The `get_fields` method omits the `server_config_` instance attribute but returns all other instance attributes:

```python
>>> server_group.get_fields()
{
    'name': 'my-server-group',
    'description': 'a collection of servers that all have the same configuration',
    'server_config': <orm.ServerConfig object at ...>,
}
```

NailGun was modeled after existing ORM layers, such as those used by Rails or Django. It's also similar to many other ORM layers such as SQLAlchemy and Storm. The problem is that traditional ORM layers all assume that you are only going to ever talk to a single database at a single time, so they let connection information be global. Robottelo's ORM layer has thus far followed the assumption, making extensive use of helpers like `get_server_url` and `get_server_credentials`. However, NailGun is trying to be a bit more powerful and flexible. Itr tries to make it possible for you to communicate with multiple servers/with different credentials/etc at the same time. Making this happen is as simple as creating a pair of entity objects and giving each one a unique server configuration. But once this is done, we've broken one of the fundamental assumptions that goes in to traditional ORM design.

That's why this hack and `get_fields` exist. It's my attempt to deal with the fact that we're not carrying around non-field instance attributes with each entity object.